# Phylogenomics course (PHYSALIA) - practical sessions for Bayesian inference
In this GitHub repository, you will find all the material that we will be using 
for the practical sessions. The content that we will cover is the following:   

   * [Day 1.a - Introduction to MCMC](practical_sessions/day1/00_R): we will use R to 
   put in practice all the theoretical concepts about Markov Chain Monte Carlo (MCMC)
   in Bayesian inference that we have gone through in the lectures. You will be able to learn how to
   write an MCMC algorithm, run an MCMC, assess the effect that using different 
   parameter values to run the MCMC can have on chain efficiency and convergence.   
   * [Day 1.b - Bayesian phylogenetic inference](practical_sessions/day1/01_MrBayes): we 
   will be using `MrBayes` to infer a phylogeny based on a Bayesian 
   approach. We will explore the effect that choosing different priors, data partitioning, or
   different nucleotide substitution models (among others) can have on phylogeny inference.   
   * [Day 2.a - Divergence times estimation (I)](practical_sessions/day2/00_BEAST): we
   will be using `BEAST` to infer the divergence times of a phylogeny inferred with virus
   sequences that were isolated at different points in time.   
   * [Day 2.b - Divergence times estimation (II)](practical_sessions/day2/01_MCMCtree): we
   will be using `MCMCtree` to infer the divergence times of a mammal phylogeny. We will see how we can
   use the approximate likelihood calculation to speed up this inference step with large genomic data 
   sets.
   
To go through the content of each practical session, just click the corresponding links provided above.

Happy inference! :) 